tools/themis/src/utils.rs

```diff
@@ -0,0 +1,135 @@
+                "{c_heading}(i) {}:{c_none}{}",
+                msg,
+                expected.map_or("".to_string(), |Expected { value, reason }| format!(
+                    " [expected: {c_expected}{}{c_none}{}]",
+                    value,
+                    reason.map_or("".to_string(), |reason| format!(", {}", reason)),
+                    c_expected = style::fg(style::Color::Color256(35))
+                        + &style::bg(style::Color::Gray { shade: 3 })
+                        + style::bold(),
+                    c_none = style::reset()
+                )),
+                c_heading = style::fg(style::Color::Color256(172)) + style::bold(),
+                c_none = style::reset()
+            );
+            let report = outliers.into_iter().fold(
+                header,
+                |mut acc, PackageOutcome { pkg: Package { parsed: pkg, .. }, value }| {
+                    acc.push_str(&format!(
+                        "\n \u{2022} {c_name}{}{c_none} v{} {c_path}({}){c_none}{}",
+                        pkg.name,
+                        pkg.version,
+                        pkg.manifest_path.strip_prefix(&workspace.root).unwrap(),
+                        value.map_or("".to_string(), |v| format!(
+                            " [found: {c_found}{}{c_none}]",
+                            v,
+                            c_found = style::fg(style::Color::White)
+                                + &style::bg(style::Color::Red)
+                                + style::bold(),
+                            c_none = style::reset()
+                        )),
+                        c_name = style::fg(style::Color::Color256(39)) + style::bold(),
+                        c_path = style::fg(style::Color::Gray { shade: 12 }),
+                        c_none = style::reset(),
+                    ));
+                    acc
+                },
+            );
+            eprintln!("{}", report);
+            Ok(true)
+        }
+        _ => Ok(false),
+    }
+}
+
+pub fn is_publishable(pkg: &Package) -> bool {
+    !matches!(pkg.raw["package"].get("publish"), Some(toml::Value::Boolean(false)))
+}
+
+macro_rules! pub_missing {
+    ($pkg:expr, $($files:tt)+) => {{
+        let pkg_root = $pkg.parsed.manifest_path.parent().unwrap();
+        $crate::utils::is_publishable($pkg) && !$crate::utils::pub_missing!(@ [pkg_root] $($files)+)
+    }};
+    (@ [$root:expr] $file:literal || $($files:tt)+) => {{
+        $crate::utils::pub_missing!(@ [$root] ($file) || $($files)+)
+    }};
+    (@ [$root:expr] ($($file:literal)&&+) $(|| $($files:tt)+)?) => {{
+        $($crate::utils::pub_missing!(@ [$root] $file))&&+ $(|| $crate::utils::pub_missing!(@ [$root] $($files)+))?
+    }};
+    (@ [$root:expr] $file:literal) => {{
+        $root.join($file).exists()
+    }};
+    (@ [$root:expr] ($file:expr)?) => {{
+        $file.as_ref().map_or(false, |file| $root.join(file).exists())
+    }};
+}
+
+pub(crate) use pub_missing;
```

Comment:
As a general readability thing, it's rater hard to figure out what this macro is supposed to do.

Reply:
True.. could do with some documentation.. or just expanding the macro in all it's invocations..

Generally, it's supposed to check for the existence of files. By default with OR, but optionally, AND:

```rust
pub_missing!(pkg, "a" || "b" || ( "c" && "d" ) || "e")
```

Checks for the existence of either `a`, `b`, `e` or both `c` and `d`..

Could also do with better naming.

Reply:
Simplified this to `exists!(pkg, <files...>)`